### PR TITLE
Added a new flag to ignore unsuccessful bind

### PR DESCRIPTION
**Pull request by sahilnaphade**
This is a feature request made here : [Issue 3056](https://github.com/patroni/patroni/issues/3056)

Ask:
A new flag to ignore bind errors in the case where user wants re-validate parameters on a running Patroni cluster. In other words, if patroni is already running, Ansible (or other orchestrators) should get a successful code; even when the ports with `listen` specified are already bound by other process (most likely another Patroni instance)

Changes discussed and made:
A variable will be set when a flag is specified in CLI. If set, this flag should ignore the port as requested.

### CODE REVIEW

File: patroni/validator.py [hunk 1]
```diff
@@ -18,6 +18,10 @@
 from .utils import data_directory_is_empty, get_major_version, parse_int, split_host_port
 
 
+# Additional parameters to fine-tune validation process
+VALIDATION_PARAMS : Dict[str, bool] = dict()
```
Comment (CyberDem0n) [inline patroni/validator.py:22]: Variable with all capital letters in its name is treated by `pyright` as a constant, which is not entirely true for this specific case.
```suggestion
_validation_params : Dict[str, Any] = {}
```

Besides that, I would prefer to have a function that will modify global variable.

File: patroni/validator.py [hunk 2]
```diff
@@ -137,7 +141,10 @@ def validate_host_port(host_port: str, listen: bool = False, multiple_hosts: boo
             s = socket.socket(proto[0][0], socket.SOCK_STREAM)
             try:
                 if s.connect_ex((host, port)) == 0:
-                    if listen:
+                    # Do not raise an exception if ignore_listen_port is set to True.
+                    if VALIDATION_PARAMS.get('ignore_listen_port', False):
+                        pass
+                    elif listen:
```
Comment (CyberDem0n) [inline patroni/validator.py:147]:
```suggestion
                    # Do not raise an exception if ignore_listen_port is set to True.
                    if listen and not _validation_params.get('ignore_listen_port', False):
```

File: patroni/__main__.py [hunk 3]
```diff
@@ -259,6 +259,8 @@ def process_arguments() -> Namespace:
                        help='Generate a Patroni yaml configuration file for a running instance')
     parser.add_argument('--dsn', help='Optional DSN string of the instance to be used as a source \
                                     for config generation. Superuser connection is required.')
+    parser.add_argument('--ignore-listen-port', '-i', action='store_true',
+                        help='Ignore if the ports in `listen` are already bound by some process.')
```
Comment (CyberDem0n) [inline patroni/__main__.py:263]: Please update the help message to make it clear that this option works only for config validation.

Comment (CyberDem0n) [reply patroni/__main__.py:263]: Besides that, please update the `docstring` of this function.

**Review by CyberDem0n — CHANGES_REQUESTED** [review]

File: patroni/__main__.py [hunk 4]
```diff
@@ -259,6 +261,9 @@ def process_arguments() -> Namespace:
                        help='Generate a Patroni yaml configuration file for a running instance')
     parser.add_argument('--dsn', help='Optional DSN string of the instance to be used as a source \
                                     for config generation. Superuser connection is required.')
+    parser.add_argument('--ignore-listen-port', '-i', action='store_true',
+                        help='Ignores unsuccessful binds for the ports marked for `listen`.\
```
Comment (hughcapet) [inline patroni/__main__.py:265]:
```suggestion
                        help='Ignore `listen` ports already in use.\
```

File: patroni/validator.py [hunk 5]
```diff
@@ -17,6 +17,19 @@
...
 from .utils import data_directory_is_empty, get_major_version, parse_int, split_host_port
 
+# Additional parameters to fine-tune validation process
+_validation_params: Dict[str, Any] = {}
+
+
+def populate_validate_params(ignore_listen_port: bool = False) -> None:
+    """Populates parameters used to fine-tune the validation of the Patroni config
+
+    :param ignore_listen_port bool: Used to ignore the bind failures for the ports marked as `listen`
```
Comment (CyberDem0n) [inline patroni/validator.py:27]:
```suggestion
    :param ignore_listen_port: ignore the bind failures for the ports marked as `listen`.
```

File: patroni/validator.py [hunk 6]
```diff
@@ -17,6 +17,19 @@
 from .log import type_logformat
 from .utils import data_directory_is_empty, get_major_version, parse_int, split_host_port
 
+# Additional parameters to fine-tune validation process
+_validation_params: Dict[str, Any] = {}
+
+
+def populate_validate_params(ignore_listen_port: bool = False) -> None:
+    """Populates parameters used to fine-tune the validation of the Patroni config
```
Comment (CyberDem0n) [inline patroni/validator.py:25]:
```suggestion
    """Populate parameters used to fine-tune the validation of the Patroni config.
```

File: patroni/__main__.py [hunk 7]
```diff
@@ -269,7 +274,8 @@ def process_arguments() -> Namespace:
         sys.exit(0)
     elif args.validate_config:
         from patroni.config import Config, ConfigParseError
-        from patroni.validator import schema
+        from patroni.validator import populate_validate_params, schema
+        populate_validate_params(ignore_listen_port=args.ignore_listen_port)
```
Comment (CyberDem0n) [inline patroni/__main__.py:278]:
```suggestion

        populate_validate_params(ignore_listen_port=args.ignore_listen_port)
```

File: patroni/__main__.py [hunk 8]
```diff
@@ -241,6 +241,8 @@ def process_arguments() -> Namespace:
       * ``--validate-config`` -- used to validate the Patroni configuration file
       * ``--generate-config`` -- used to generate Patroni configuration from a running PostgreSQL instance
       * ``--generate-sample-config`` -- used to generate a sample Patroni configuration
+      * ``--ignore-listen-port`` | ``-i`` -- used to ignore unsuccessful binds to the ports marked with `listen`.\
```
Comment (hughcapet) [inline patroni/__main__.py:244]:
```suggestion
      * ``--ignore-listen-port`` | ``-i`` -- used to ignore ``listen`` ports already in use.\
```

File: patroni/__main__.py [hunk 9]
```diff
@@ -241,6 +241,8 @@ def process_arguments() -> Namespace:
...
+          Can be used only with --validate-config
```
Comment (hughcapet) [inline patroni/__main__.py:245]:
```suggestion
          Can be used only with ``--validate-config``
```

**Comment by hughcapet** [comment]
we also need to extend docs
https://github.com/patroni/patroni/blob/835d93951d437cacc24e3925272559cb20267c4d/docs/patroni_configuration.rst?plain=1#L241

**Comment by sahilnaphade** [comment]
Hi @CyberDem0n and @hughcapet 
Thanks for your patience in reviewing my PR. I have updated the documentations as requested. Please let me know if anything else requires addressing.

File: patroni/__main__.py [hunk 10]
```diff
@@ -241,6 +241,8 @@ def process_arguments() -> Namespace:
       * ``--validate-config`` -- used to validate the Patroni configuration file
       * ``--generate-config`` -- used to generate Patroni configuration from a running PostgreSQL instance
       * ``--generate-sample-config`` -- used to generate a sample Patroni configuration
+      * ``--ignore-listen-port`` | ``-i`` -- used to ignore ``listen`` ports already in use.\
```
Comment (hughcapet) [inline patroni/__main__.py:244]:
```suggestion
      * ``--ignore-listen-port`` | ``-i`` -- used to ignore ``listen`` ports already in use.
```

**Review by hughcapet — APPROVED** [review]

**Review by CyberDem0n — APPROVED** [review]

**Pull request merged**